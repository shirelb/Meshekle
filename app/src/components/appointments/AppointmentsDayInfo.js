import React, {Component} from 'react';
import {Modal, ScrollView, StyleSheet, Text} from 'react-native';
import moment from 'moment';
import {List} from 'react-native-paper';
import Button from '../submitButton/Button';
import mappers from "../../shared/mappers";


export default class AppointmentsDayInfo extends Component {
    constructor(props) {
        super(props);

        this.state = {
            dateModalVisible: this.props.dateModalVisible,
            selectedDate: this.props.selectedDate,
            markedDates: this.props.markedDates,
            expanded: this.props.expanded,
        };

    }

    componentWillReceiveProps(nextProps, nextContext) {
        console.log('nextProps. ', nextProps);

        this.setState({
            dateModalVisible: nextProps.dateModalVisible,
            selectedDate: nextProps.selectedDate,
            markedDates: nextProps.markedDates,
            expanded: nextProps.expanded,
        })
    }


    render() {

        return (
            <Modal
                animationType="fade"
                transparent={false}
                visible={this.state.dateModalVisible}
                onRequestClose={() => {
                    this.setState({dateModalVisible: false});
                    this.props.afterCloseModalShowSelectDay();
                }}
            >
                <ScrollView style={{marginTop: 22}}>
                    <ScrollView>
                        <Text style={styles.textTitle}>{this.state.selectedDate}</Text>

                        <Button
                            label='חזור'
                            onPress={() => {
                                this.setState({dateModalVisible: false})
                                this.props.afterCloseModalShowSelectDay();
                            }}
                        />

                        <Button
                            label='בקש תור חדש'
                            onPress={() => {
                                this.props.onAppointmentRequestPress(this.state.selectedDate);
                                this.setState({dateModalVisible: false});
                            }}
                        />

                        <List.Section title={'תורים'}>
                            {this.state.selectedDate === '' || this.state.markedDates[this.state.selectedDate].appointments.length === 0 ?
                                <Text>אין תורים לתאריך זה</Text>
                                :
                                this.state.markedDates[this.state.selectedDate].appointments.map(item => {
                                    return <List.Accordion
                                        key={item.appointmentId}
                                        title={moment(item.startDateAndTime).format('HH:mm') + '-' + moment(item.endDateAndTime).format('HH:mm')}
                                        description={mappers.serviceProviderRolesMapper(item.AppointmentDetail.role) + ' - ' + item.serviceProviderFullname}
                                        left={props => <List.Icon {...props} icon="perm-contact-calendar"/>}
                                        expanded={this.state.expanded[item.appointmentId]}
                                        onPress={() => {
                                            let expanded = this.state.expanded;
                                            expanded[item.appointmentId] = !expanded[item.appointmentId];
                                            this.setState({expanded: expanded})
                                        }}
                                    >
                                        <List.Item
                                            title={JSON.parse(item.AppointmentDetail.subject).join(", ")}
                                            description={item.remarks}
                                            containerStyle={{borderBottomWidth: 0}}
                                        />
                                    </List.Accordion>
                                })
                            }
                        </List.Section>
                    </ScrollView>
                </ScrollView>
            </Modal>
        );
    }
}

const styles = StyleSheet.create({
    textTitle: {
        textAlign: 'center',
        color: '#050505',
        fontWeight: 'bold',
        fontSize: 20,
        marginBottom: 22
    }
});
