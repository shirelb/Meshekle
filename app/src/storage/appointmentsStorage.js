import axios from "axios";
import {APP_SOCKET, SERVER_URL} from "../shared/constants";


var getUserAppointmentRequests = function (userId, userHeaders) {
    return axios.get(`${SERVER_URL}/api/appointmentRequests/user/userId/${userId}`,
        {
            headers: userHeaders,
            params: {
                status: 'requested'
            },
        })
        .then(response => {
            return response;
        })
        .catch(error => {
            console.log('get user appointment requests error ', error)
        });
};

var getUserAppointments = function (userId, userHeaders) {
    return axios.get(`${SERVER_URL}/api/appointments/user/userId/${userId}`,
        {
            headers: userHeaders,
            params: {
                status: 'set'
            },
        })
        .then(response => {
            return response;
        })
        .catch(error => {
            console.log('get user appointments error ', error)
        });
};

var getUserAppointmentById = function (userId, userHeaders, eventId) {
    return axios.get(`${SERVER_URL}/api/appointments/user/userId/${userId}`,
        {
            headers: userHeaders,
            params: {status: 'set', appointmentId: eventId}
        })
        .then(response => {
            return response;
        })
        .catch(error => {
            console.log('get user appointments by id error ', error)
        });
};

var postUserAppointmentRequest = function (userId, serviceProvider, appointmentRequest, userHeaders) {
    return axios.post(`${SERVER_URL}/api/appointmentRequests/user/request`,
        {
            userId: userId,
            serviceProviderId: serviceProvider.serviceProviderId,
            role: serviceProvider.role,
            availableTime: appointmentRequest.availableTime,
            notes: appointmentRequest.notes ? appointmentRequest.notes : '',
            subject: JSON.stringify(appointmentRequest.subject)
        },
        {
            headers: userHeaders,
        })
        .then(response => {
            APP_SOCKET.emit('userPostAppointmentRequests', {
                serviceProviderId: serviceProvider.serviceProviderId
            });

            return response;
        })
        .catch(error => {
            console.log('post user appointment request error ', error)
        });
};

var cancelAppointmentRequestById = (appointmentRequest, headers) => {
    return axios.put(`${SERVER_URL}/api/appointmentRequests/user/reject`,
        {
            userId: appointmentRequest.AppointmentDetail.clientId,
            appointmentRequestId: appointmentRequest.requestId,
        },
        {
            headers: headers,
            params: {
                status: 'rejected'
            },
        }
    )
        .then((response) => {
            APP_SOCKET.emit('userCancelAppointmentRequests', {
                serviceProviderId: appointmentRequest.AppointmentDetail.serviceProviderId,
            });

            return response
        })
        .catch((error) => {
            console.log('reject appointment request error ', error);
        });
};

var cancelAppointmentById = (appointment, headers) => {
    return axios.put(`${SERVER_URL}/api/appointments/user/cancel/userId/${appointment.AppointmentDetail.clientId}/appointmentId/${appointment.appointmentId}`,
        {},
        {
            headers: headers
        }
    )
        .then((response) => {
            APP_SOCKET.emit('userCancelAppointment', {
                serviceProviderId: appointment.AppointmentDetail.serviceProviderId,
            });

            return response
        })
        .catch((error) => {
            console.log('cancel appointment error ', error);
        });
};


export default {
    getUserAppointments,
    getUserAppointmentById,
    postUserAppointmentRequest,
    getUserAppointmentRequests,
    cancelAppointmentRequestById,
    cancelAppointmentById,
};
