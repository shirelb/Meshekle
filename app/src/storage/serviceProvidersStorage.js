import axios from "axios";
import {SERVER_URL} from "../shared/constants";

var getServiceProviders = function (userHeaders) {
    return axios.get(`${SERVER_URL}/api/serviceProviders`,
        {headers: userHeaders}
    )
        .then(response => {
            return response;
        })
        .catch(error => {
            console.log('load appointments error ', error)
        });
};


export default {getServiceProviders};
