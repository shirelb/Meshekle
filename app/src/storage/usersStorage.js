import axios from "axios";
import {SERVER_URL} from "../shared/constants";

var getUserEvents = function (userId, userHeaders) {
    return axios.get(`${SERVER_URL}/api/users/events/userId/${userId}`,
        {headers: userHeaders}
    )
        .then(response => {
            return response;
        })
        .catch(error => {
            console.log('get user events error ', error)
        });
};

var getUserById = function (userId, userHeaders) {
    return axios.get(`${SERVER_URL}/api/users/userId/${userId}`,
        {headers: userHeaders}
    )
        .then(response => {
            return response;
        })
        .catch(error => {
            console.log('get user by id error ', error)
        });
};

var getUsers = function (userHeaders) {
    return axios.get(`${SERVER_URL}/api/users`,
        {headers: userHeaders}
    )
        .then(response => {
            return response.data.filter(user => user.userId !== 1 && user.userId !== "1");
        })
        .catch(error => {
            console.log('get user by id error ', error)
        });
};

var userValidToken = function (token) {
    return axios.post(`${SERVER_URL}/api/users/validToken`,
        {
            "token": token,
        },
    );
}

var userLogin = function (userId, password) {
    return axios.post(`${SERVER_URL}/api/users/login/authenticate`,
        {
            "userId": userId,
            "password": password
        },
    );
}


export default {getUserEvents, getUserById, userValidToken, userLogin, getUsers};
