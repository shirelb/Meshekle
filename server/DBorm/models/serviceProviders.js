module.exports = (sequelize, type) => {
    return sequelize.define('ServiceProviders', {
        serviceProviderId:{
            type: type.INTEGER,
            primaryKey: true,
            autoIncrement: true,
            allowNull: false,
            required: true
        },
        role : {
            type: type.STRING,
            allowNull: false,
            required: true
        },
        userId: {
            type: type.INTEGER,
            allowNull: false,
            required: true
        },
        operationTime: {
            type: type.STRING,
            allowNull: false,
            required: true
        },
        phoneNumber: {
            type: type.STRING,
            allowNull: false,
            required: true
        },
    });
};