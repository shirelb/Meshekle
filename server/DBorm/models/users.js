module.exports = (sequelize, type) => {
    return sequelize.define('Users', {
        userId:{
            type: type.STRING,
            primaryKey: true,
            required: true
        },
        fullname: {
            type: type.STRING,
            required: true
        },
        password: {
            type: type.STRING,
            required: true
        },
        email: {
            type: type.STRING,
            required: true
        },
        mailbox: {
            type: type.INTEGER,
            required: true
        },
        cellphone: {
            type: type.STRING,
            required: true
        },
        phone: {
            type: type.STRING
        },
        bornDate: {
            type: type.DATE,
            required: true
        },
        active: {
            type: type.BOOLEAN,
            required: true,
            defaultValue: true
        },
        image: {
            type: type.BLOB,
        }
    });
};