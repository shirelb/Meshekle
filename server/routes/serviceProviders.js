var express = require('express');
var router = express.Router();
const Sequelize = require('sequelize');
const Op = Sequelize.Op;
const {Users, ServiceProviders, AppointmentRequests, Appointments, Incidents, Events} = require('../DBorm/DBorm');

/* GET users listing. */
router.get('/', function(req, res, next) {
  res.send('respond with a resource serviceProviders');
});


/* POST user . */
router.post('/users/add', function (req, res, next) {
    Users.create({
        userId: req.body.userId,
        fullname: req.body.fullname,
        password: req.body.password,
        email: req.body.email,
        mailbox: req.body.mailbox,
        cellphone: req.body.cellphone,
        phone: req.body.phone,
        bornDate: new Date(req.body.bornDate),
        active: true,
    })
        .then((newUser) => {
            console.log('New user' + newUser.username + ', with id ' + newUser.userId + ' has been created.');
            res.status(200).send({"message": "User successfully added!", newUser});
        })
        .catch(err => {
            console.log(err);
            res.status(500).send(err);
        })
});

/* POST user . */
router.post('/add', function (req, res, next) {
    ServiceProviders.create({
        serviceProviderId: req.body.serviceProviderId,
        userId: req.body.userId,
        role: req.body.role,
        operationTime: req.body.operationTime,
        phoneNumber: req.body.phoneNumber,
        appointmentWayType: req.body.appointmentWayType,
        active: true,
    })
        .then((newUser) => {
            console.log('New user' + newUser.username + ', with id ' + newUser.userId + ' has been created.');
            res.status(200).send({"message": "User successfully added!", newUser});
        })
        .catch(err => {
            console.log(err);
            res.status(500).send(err);
        })
});

module.exports = router;
