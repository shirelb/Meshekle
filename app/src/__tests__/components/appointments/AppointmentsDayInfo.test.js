import React from 'react';

import {Modal} from "react-native";
import {List} from "react-native-paper";
import {shallow} from "enzyme/build";
import AppointmentsDayInfo from "../../../components/appointments/AppointmentsDayInfo";
import usersStorage from "../../../storage/usersStorage";
import phoneStorage from "react-native-simple-store";
import users from "../../jsons/users";
import serviceProvidersStorage from "../../../storage/serviceProvidersStorage";
import serviceProviders from "../../jsons/serviceProviders";
import appointmentsStorage from "../../../storage/appointmentsStorage";
import user013637605AppointmentRequests from "../../jsons/user013637605AppointmentRequests";

jest.mock("../../../storage/serviceProvidersStorage");
jest.mock("../../../storage/usersStorage");
jest.mock("react-native-simple-store");


describe("AppointmentsDayInfo should", () => {
    let wrapper = null;
    let componentInstance = null;

    const props = {
        dateModalVisible: true,
        selectedDate: "2019-07-02",
        markedDates: {
            "2019-05-05": {
                "marked": true,
                "selected": false,
                "appointments": [
                    {
                        "appointmentId": 57,
                        "startDateAndTime": "2019-05-04T22:33:13.000Z",
                        "endDateAndTime": "2019-05-05T01:26:35.000Z",
                        "remarks": "a odio in hac habitasse platea",
                        "status": "set",
                        "createdAt": "2019-06-03T20:41:32.581Z",
                        "updatedAt": "2019-06-03T20:41:32.581Z",
                        "AppointmentDetail": {
                            "appointmentId": 57,
                            "clientId": "052957237",
                            "serviceProviderId": 468254910,
                            "role": "appointmentsDentist",
                            "subject": "[\"שיננית\", \"טיפול שורש\", \"עקירה\", \"יישור שיניים\"]",
                            "createdAt": "2019-06-03T20:41:32.543Z",
                            "updatedAt": "2019-06-03T20:41:32.543Z",
                            "userId": null
                        }
                    }
                ]
            },
            "2019-07-02": {
                "marked": true,
                "selected": true,
                "appointments": [
                    {
                        "appointmentId": 223,
                        "startDateAndTime": "2019-07-02T16:03:23.000Z",
                        "endDateAndTime": "2019-07-02T01:34:14.000Z",
                        "remarks": "sapien quis libero nullam sit amet turpis elementum",
                        "status": "set",
                        "createdAt": "2019-06-03T20:41:32.581Z",
                        "updatedAt": "2019-06-03T20:41:32.581Z",
                        "AppointmentDetail": {
                            "appointmentId": 223,
                            "clientId": "052957237",
                            "serviceProviderId": 229971094,
                            "role": "appointmentsDentist",
                            "subject": "[\"יישור שיניים\", \"ניקוי\", \"עקירה\", \"טיפול שורש\"]",
                            "createdAt": "2019-06-03T20:41:32.543Z",
                            "updatedAt": "2019-06-03T20:41:32.543Z",
                            "userId": null
                        },
                        "serviceProviderFullname": "Ronda Cail"
                    }
                ],
                "color": "blue"
            }
        },
        expanded: false,
        loadAppointments: jest.fn(),
        afterCloseModalShowSelectDay: jest.fn(),
        onAppointmentRequestPress: jest.fn(),
    };
    const mockStore = {
        userData: {
            serviceProviderId: "549963652",
            userId: "549963652",
            token: "some token"
        }
    };
    const navigation = {navigate: jest.fn(), dispatch: jest.fn()};

    phoneStorage.get = jest.fn().mockImplementation((key) => Promise.resolve(mockStore[key]));
    phoneStorage.update = jest.fn().mockImplementation((key, value) => Promise.resolve(mockStore[key] = value));
    usersStorage.getUsers = jest.fn().mockImplementation(() => Promise.resolve(users));
    usersStorage.getUserByUserID = jest.fn().mockImplementation((userId) => Promise.resolve(users.filter(user => user.userId === userId)[0]));
    serviceProvidersStorage.getServiceProviders = jest.fn().mockResolvedValue(serviceProviders);
    serviceProvidersStorage.getServiceProviderUserDetails = jest.fn().mockImplementation((serviceProviderId) => Promise.resolve({data: users.filter(user => user.userId === serviceProviderId)[0]}));
    serviceProvidersStorage.getServiceProviderById = jest.fn().mockImplementation((serviceProviderId) => Promise.resolve(serviceProviders.filter(provider =>
        provider.serviceProviderId === serviceProviderId)));
    appointmentsStorage.getUserAppointmentRequests = jest.fn().mockImplementation(() => Promise.resolve(user013637605AppointmentRequests));
    appointmentsStorage.postUserAppointmentRequest = jest.fn().mockImplementation(() => Promise.resolve());
    appointmentsStorage.cancelAppointmentById = jest.fn().mockImplementation(() => Promise.resolve());

    beforeEach(() => {
        jest.clearAllMocks();
    });

    it('match snapshot', async () => {
        wrapper = shallow(<AppointmentsDayInfo {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("render what the user see", async () => {
        wrapper = await shallow(<AppointmentsDayInfo {...props} />);
        componentInstance = wrapper.instance();

        await wrapper.update();

        expect(wrapper.find(Modal)).toHaveLength(1);
        expect(wrapper.find(List.Section)).toHaveLength(1);
        expect(wrapper.find(List.Accordion)).toHaveLength(1);
        expect(wrapper.find(List.Item)).toHaveLength(1);
        expect(wrapper.find('Button')).toHaveLength(2);
        expect(wrapper.find('Button').at(0).props().label).toEqual('חזור');
        expect(wrapper.find('Button').at(1).props().label).toEqual('בקש תור חדש');
    });

    it("mounted with the right data", async () => {
        expect(componentInstance.state.dateModalVisible).toEqual(true);
        expect(componentInstance.state.selectedDate).toEqual("2019-07-02");
    });

    it("handle cancelAppointment", async () => {
        wrapper = shallow(<AppointmentsDayInfo {...props} />);
        componentInstance = wrapper.instance();

        await componentInstance.cancelAppointment();
        expect(appointmentsStorage.cancelAppointmentById).toHaveBeenCalled();
    });

});