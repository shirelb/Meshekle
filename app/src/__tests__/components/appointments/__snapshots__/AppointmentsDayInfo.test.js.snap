// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppointmentsDayInfo should match snapshot 1`] = `
<Component
  animationType="fade"
  hardwareAccelerated={false}
  onRequestClose={[Function]}
  transparent={false}
  visible={true}
>
  <ScrollViewMock
    style={
      Object {
        "marginTop": 22,
      }
    }
  >
    <ScrollViewMock>
      <Text
        style={
          Object {
            "color": "#050505",
            "fontSize": 20,
            "fontWeight": "bold",
            "marginBottom": 22,
            "textAlign": "center",
          }
        }
      >
        2019-07-02
      </Text>
      <Button
        label="חזור"
        onPress={[Function]}
      />
      <Button
        label="בקש תור חדש"
        onPress={[Function]}
      />
      <withTheme(List.Section)
        style={
          Object {
            "textAlign": "right",
          }
        }
        title="תורים"
      >
        <View
          key="223"
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "flex-start",
            }
          }
        >
          <View
            style={
              Object {
                "width": "20%",
              }
            }
          >
            <Icon
              color="red"
              name="delete-forever"
              onPress={[Function]}
              raised={false}
              reverse={false}
              reverseColor="white"
              size={24}
              underlayColor="white"
            />
          </View>
          <View
            style={
              Object {
                "width": "80%",
              }
            }
          >
            <withTheme(List.Accordion)
              description="מרפאת שיניים - Ronda Cail"
              onPress={[Function]}
              title="19:03-04:34"
            >
              <withTheme(List.Item)
                containerStyle={
                  Object {
                    "borderBottomWidth": 0,
                  }
                }
                description="sapien quis libero nullam sit amet turpis elementum"
                title="יישור שיניים, ניקוי, עקירה, טיפול שורש"
              />
            </withTheme(List.Accordion)>
          </View>
        </View>
      </withTheme(List.Section)>
    </ScrollViewMock>
  </ScrollViewMock>
</Component>
`;
